pac4j-http/src/main/java/org/pac4j/http/credentials/DigestCredentials.java

```diff
@@ -0,0 +1,136 @@
+import org.pac4j.core.util.CommonHelper;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.StringTokenizer;
+
+/**
+ * <p>This credentials is retrieved from a HTTP request.</p>
+ * <p>A user profile can be attached with the credentials if it has been created by a {@link org.pac4j.http.credentials.authenticator.Authenticator}.
+ * In that case, the {@link org.pac4j.http.profile.creator.AuthenticatorProfileCreator} must be used to retrieve the attached user profile.</p>
+ *
+ * @author Mircea Carasel
+ */
+public class DigestCredentials extends TokenCredentials {
+
+    private String username;
+    private String response;
+
+    private String realm;
+    private String nonce;
+    private String uri;
+    private String cnonce;
+    private String nc;
+    private String qop;
+
+    private String httpMethod;
+
+    public DigestCredentials(final String token, final String httpMethod, final String clientName) {
+        super(token, clientName);
+        Map<String, String> valueMap = parseTokenValue(token);
+
+        username = valueMap.get("username");
+        response = valueMap.get("response");
+
+        if (CommonHelper.isBlank(username) || CommonHelper.isBlank(response)) {
+            throw new CredentialsException("Bad format of the digest auth header");
+        }
+        realm = valueMap.get("realm");
+        nonce = valueMap.get("nonce");
+        uri = valueMap.get("uri");
+        cnonce = valueMap.get("cnonce");
+        nc = valueMap.get("nc");
+        qop = valueMap.get("qop");
+        this.httpMethod = httpMethod;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public String getResponse() {
+        return response;
+    }
+
+    public String calculateServerDigest(boolean passwordAlreadyEncoded, String password) {
+        return generateDigest(passwordAlreadyEncoded, username,
+                realm, password, httpMethod, uri, qop, nonce, nc, cnonce);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        DigestCredentials that = (DigestCredentials) o;
+
+        if (username != null ? !username.equals(that.username) : that.username != null) return false;
+        return !(response != null ? !response.equals(that.response) : that.response != null);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = username != null ? username.hashCode() : 0;
+        result = 31 * result + (response != null ? response.hashCode() : 0);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return CommonHelper.toString(this.getClass(), "username", this.username, "response", "[PROTECTED]",
+                "clientName", getClientName());
+    }
+
+    private Map<String, String> parseTokenValue(String token) {
```

Comment:
Basically what follows after "Digest " in the Authorization HTTP header is a token composed of different values like: nonce, nc, cnonce, qop, response...this is why I thought that DigestCredentials should extend TokenCredentials

Further, the DigestAuthExtractor retrieves the generic TokenCredentials from the Authorization header, just like in BasicAuthExtractor. then it converts it to DigestCredentials (token parsing incapsulated in DigetCredentials). 

I can change this so DigestCredential to extend HttpCredentials and enhance token parsing inside DigestAuthExtractor. What do you think ? thanks!